Python/ceval.c

```diff
@@ -857,6 +857,32 @@ static const binaryfunc binary_ops[] = {
     [NB_INPLACE_XOR] = PyNumber_InPlaceXor,
 };
 
+#define BINARY_OP_FAST_FLOAT(OP)                       \
+    do {                                               \
+        assert(cframe.use_tracing == 0);               \
+        PyObject *lhs = SECOND();                      \
+        PyObject *rhs = TOP();                         \
+        DEOPT_IF(!PyFloat_CheckExact(lhs), BINARY_OP); \
+        DEOPT_IF(!PyFloat_CheckExact(rhs), BINARY_OP); \
+        STAT_INC(BINARY_OP, hit);                      \
+        double l = PyFloat_AS_DOUBLE(lhs);             \
+        double r = PyFloat_AS_DOUBLE(rhs);             \
+        double d = l OP r;                             \
+        Py_DECREF(rhs);                                \
+        STACK_SHRINK(1);                               \
+        if (Py_REFCNT(lhs) == 1) {                     \
```

Comment:
Would there be any benefit in checking whether perhaps rhs has a refcount of 1, in case lhs doesn't? While people are probably more likely to write `return x+1` rather than `return 1+x` (though even I could sometimes see a reason to do that, from "rhyming" with some other operations or just how the textbook presents the formula), writing `return 2*x` is more likely than `return x*2`. And of course for asymmetric operations you may not have a choice, like `return 1/x`.

Of course, there's a cost (more code, more branches). There's also the fact that `+=` and friends steer this in the direction you're taking here.

Reply:
> Of course, there's a cost (more code, more branches). There's also the fact that `+=` and friends steer this in the direction you're taking here.

Yeah, those are my main motivations for sticking with the LHS here.